media/open_zim.ksy

```diff
@@ -0,0 +1,356 @@
+        enums:
+          namespace:
+            0x2d: #'-'
+              id: layout
+              doc: eg. the LayoutPage, CSS, favicon.png (48x48), JavaScript and images not related to the articles
+            0x41: #'A'
+              id: article
+              doc: articles - see Article Format
+            0x42: #'B'
+              id: article_meta_data
+              doc: article meta data - see Article Format
+            0x49: #'I'
+              id: image_file
+              doc: images, files - see Image Handling
+            0x4A: #'J'
+              id: image_text
+              doc: images, text - see Image Handling
+            0x4d: #'M'
+              id: meta_data
+              doc: ZIM metadata - see Metadata
+            0x55: #'U'
+              id: category_text
+              doc: categories, text - see Category Handling
+            0x56: #'V'
+              id: category_article_list
+              doc: categories, article list - see Category Handling
+            0x57: #'W'
+              id: category_list_per_article
+              doc: categories per article, category list - see Category Handling
+            0x58: #'X'
+              id: fulltext_index
+              doc: fulltext index - see ZIM Index Format
+  cluster_ptr:
+    seq:
+      - id: ptr
+        type: u8
+    instances:
+      cluster:
+        pos: ptr
+        type: cluster(ptr)
+    types:
+      cluster:
+        # BUG: JUNK. Seems to be positioned to a wrong offsets. info.reserved is not 0
+        params:
+          - id: ptr
+            type: u8
+        seq:
+          - id: info
+            type: info_t
+          - id: blobs_ptrs_compressed
+            process: kaitai.compress.lzma(2)
+            size: info.ptr_size * blob_count_plus_one
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::lzma2
+          - id: blobs_ptrs_uncompressed
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::none or info.compression == compression::default
+        instances:
+          first_blob_offset:
+            value: ptr + 1 # info_t.size
+          blobs: 
+            type: blob(_index)
+            repeat: expr
+            repeat-expr: blob_count
+          first_blob_ptr_compressed:
+            pos: first_blob_offset
+            process: kaitai.compress.lzma(2)
+            size: 8 + _root.minimal_xz_lzma_size # max(8, 4) + minimal_xz_lzma_size
+            type: blob_ptr_u
+            if: info.compression == compression::lzma2
+          first_blob_ptr_uncompressed:
+            pos: first_blob_offset
+            type: blob_ptr_u
+            if: info.compression == compression::none or info.compression == compression::default
+          first_blob_ptr_u_v:
+            value: (info.compression == compression::lzma2?first_blob_ptr_compressed:first_blob_ptr_uncompressed).as<blob_ptr_u>.value
```

Comment:
Beware that `first_blob_ptr_uncompressed` you reference in case of `info.compression != compression::lzma2` is an optional value, which has a meaningful value only `if: info.compression == compression::none or info.compression == compression::default`.

However, {A}: `info.compression != compression::lzma2` is not exactly equivalent to {B}: `info.compression == compression::none or info.compression == compression::default`, so doing `.as<blob_ptr_u>.value` will trigger a `NullPointerException` if `{A} == true and {B} == false`.

The solution is to restrict the [attribute _domain_](https://en.wikipedia.org/wiki/Domain_of_a_function) (respectively the conditions under which the instance has a value) simply using `if`:

```suggestion
          first_blob_ptr_u_v:
            value: (info.compression == compression::lzma2?first_blob_ptr_compressed:first_blob_ptr_uncompressed).as<blob_ptr_u>.value
            if: |
              info.compression == compression::lzma2
              or info.compression == compression::none
              or info.compression == compression::default
```

Reply:
I guess

```yaml
               - id: compression
                type: b4
                enum: compression
                valid:
                  expr: _ != compression::bzip2 and _ != compression::zlib
```

would do the trick better.